pkg/scheduler/scheduling_application.go

```diff
@@ -418,12 +431,34 @@ func (sa *SchedulingApplication) tryAllocate(headRoom *resources.Resource, ctx *
 	for _, request := range sa.sortedRequests {
 		// resource must fit in headroom otherwise skip the request
 		if !resources.FitIn(headRoom, request.AllocatedResource) {
+			// if the queue (or any of its parent) has max capacity is defined,
+			// get the max headroom, this represents the configured queue quota.
+			// if queue quota is enough, but headroom is not, usually this means
+			// the cluster needs to scale up to meet the its capacity.
+			maxHeadRoom := sa.queue.getMaxHeadRoom()
+			if resources.FitIn(maxHeadRoom, request.AllocatedResource) {
+				sa.updateContainerSchedulingStateIfNeeded(request,
+					si.UpdateContainerSchedulingStateRequest_FAILED,
+					"failed to schedule the request because partition resource is not enough")
+			}
+			// skip the request
 			continue
 		}
 		if nodeIterator := ctx.getNodeIterator(); nodeIterator != nil {
 			alloc := sa.tryNodes(request, nodeIterator)
-			// have a candidate return it
-			if alloc != nil {
+			if alloc == nil {
+				// we have enough headroom, but we could not find a node for this request,
+				// this can happen when non of the nodes is qualified for this request,
+				// by satisfying both conditions:
+				//   1) node has enough resources;
+				//   2) node satisfies all placement constraints of the request (e.g predicates)
```

Comment:
I think it is working like this:
1. if we visit all nodes, no allocation or reservation is found (alloc==nil), then we do trigger the auto-scaling for this pod.
2. if we have a reservation, we hold on triggering the auto-scaling for it; but we will need to revisit the request again, that will hit  -> 1 again. this way we still need to trigger the auto-scaling for it.

Reply:
In this case we might not have been able to allocate or reserve due to affinity. Scaling up a node will not help. This is the case that in earlier discussions we tried to prevent from causing the cluster to scale up.

In a smaller cluster we might have already reserved all nodes and thus return no allocation. These reservations would have already triggered scale up.
I think we need to be conservative in this case and scale up on reservation:
```
	if alloc.result == reserved {
		sa.updateContainerSchedulingState(...)
	}
	return alloc
```

Reply:
There are a few issues if we only trigger auto-scale based on reservation:
- Compute nodes might be 0, that means no reservation at all at the initial phase
- On a small cluster, this approach will dramatically slow down the pace of auto-scaling. If we have lots of pods pending, it may only trigger the auto-scaling 1 node at a time.

Reply:
The scale up with 0 nodes will not be triggered here at all. It will have been triggered above in the `FitIn()` check. The `headRoom` check fails and the `maxHeadRoom` passes. We thus scale immediately. We cannot reserve in that case or even try to allocate if there are no nodes.

In small clusters we also do not want to go overboard and keep scaling up. If there is just one node, or maybe a couple of nodes, in the cluster it is also better to not go and trigger large numbers of scale up. The scaled up node will more often than not fit a number of requests. That should thus account for the backlog. 
If we still run out we keep scaling up. Keep in mind that we also scale up if the container does not fit in the headroom. So for large container we might have already triggered scale up due to headroom restrictions.

Reply:
this is no longer valid with the latest changes.